plutus-tx/src/PlutusTx/Blueprint/Validator.hs

```diff
@@ -0,0 +1,58 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE OverloadedStrings  #-}
+{-# LANGUAGE RecordWildCards    #-}
+
+module PlutusTx.Blueprint.Validator where
+
+import Prelude
+
+import Data.Aeson (ToJSON (..), (.=))
+import Data.Aeson qualified as Aeson
+import Data.ByteString (ByteString)
+import Data.ByteString.Base16 qualified as Base16
+import Data.List.NonEmpty (NonEmpty)
+import Data.Maybe (catMaybes)
+import Data.Text (Text)
+import Data.Text.Encoding qualified as Text
+import PlutusTx.Blueprint.Argument (ArgumentBlueprint)
+import PlutusTx.Blueprint.Parameter (ParameterBlueprint)
+
+data ValidatorBlueprint = MkValidatorBlueprint
+  { validatorTitle        :: Text
+  -- ^ A short and descriptive name for the validator.
+  , validatorDescription  :: Maybe Text
+  -- ^ An informative description of the validator.
+  , validatorRedeemer     :: ArgumentBlueprint
+  -- ^ A description of the redeemer format expected by this validator.
+  , validatorDatum        :: Maybe ArgumentBlueprint
+  -- ^ A description of the datum format expected by this validator.
+  , validatorParameters   :: Maybe (NonEmpty ParameterBlueprint)
+  -- ^ A list of parameters required by the script.
+  , validatorCompiledCode :: Maybe CompiledCode
+  -- ^ A compiled contract code with its hash.
+  }
+  deriving stock (Show)
+
+instance ToJSON ValidatorBlueprint where
+  toJSON MkValidatorBlueprint{..} =
+    Aeson.object
+      $ catMaybes
+        [ Just $ "title" .= validatorTitle
+        , ("description" .=) <$> validatorDescription
+        , Just $ "redeemer" .= validatorRedeemer
+        , ("datum" .=) <$> validatorDatum
+        , ("parameters" .=) <$> validatorParameters
+        , ("compiledCode" .=) . toHex . compiledCodeBytes <$> validatorCompiledCode
+        , ("hash" .=) . toHex . compiledCodeHash <$> validatorCompiledCode
+        ]
+   where
+    toHex :: ByteString -> Text
+    toHex = Text.decodeUtf8 . Base16.encode
```

Comment:
I always panic when text encodings make  an appearance,  but I did some experiments in ghci and this looks as if it's doing the right thing (going by the example in the CIP).  However, all that the CIP says about `compiledCode`  is that it should contain "The full compiled and cbor-encoded serialized flat script".  The compiled script is a binary object, so it has to be converted into some textual form for inclusion in JSON and I don't think they ever say exactly how that conversion is done (although maybe there's some standard JSON way of doing this that I'm not aware of).  Perhaps we should ask the CIP authors to be more explicit about the format they expect.

Reply:
Agree